### ISSUE TITLE
/dev/loop0 volume maxed out upon docker image pulls

### ISSUE THREAD

**Issue by ghost** [concourse/concourse#544]
When concourse is running a job, pulling a docker image for a custom resource_type, it fails after 3 attempts complaining that there's no space left on the device. When this error occurs, the only maxed out mount is the `/dev/loop0` device which supports the path to which the worker `--work-dir` is pointing.

The relevant mounts are:

```
Filesystem                  Size  Used Avail Use% Mounted on
/dev/xvda1                  8.0G  2.0G  6.1G  25% /
/dev/mapper/concourse-data   99G   11G   88G  11% /opt/company
/dev/loop0                  8.0G  3.5G  3.5G  50% /opt/company/concourse/worker/volumes
```

Of note:
- Running on AWS
- Using the binary distribution and manual deployment (not BOSH)
- The concourse binaries are installed in /opt/company/concourse
- `/dev/mapper/concourse-data` is a 100GB SSD based EBS volume
- Both the ATC and a single worker are running on this same system (for now)
- The worker is executed with `--work-dir=/opt/company/concourse/worker`
- The loopback mount appears to allocate space equivalent to that of the root file system size, not the larger `/opt/company` volume.
- Some runs of the same job succeed, but eventually fill up the work dir mount then fail.
- The used/available sizes pasted above do not reflect the error condition; Use is 100% on /dev/loop0 when it fails.

**Comment by vito** [comment]
Sounds like the root of the problem is that it sized the loopback image incorrectly.

Here's the bit that figures out the size:

https://github.com/concourse/bin/blob/master/cmd/concourse/worker_linux.go#L156-L170

It does it by creating the `volumes` dir and `statfs`ing it to figure out the filesystem size. Not sure why that wouldn't be working.

You could try formatting the whole 100GB disk with `btrfs` in which case it'll skip the loopback setup entirely. But I'd like to know why that code ain't working.

**Comment by troyappcarousel** [comment]
Converting to btrfs avoided the loopback volume, as you suggested, and bypassed the issue.

**Comment by vito** [comment]
Seems to correctly size the loopback for me. Spun up an EC2 instance with the same characteristics:

```
ubuntu@ip-172-31-33-124:~$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.9G   12K  3.9G   1% /dev
tmpfs           799M  348K  799M   1% /run
/dev/xvda1      7.8G  1.1G  6.3G  15% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
none            5.0M     0  5.0M   0% /run/lock
none            3.9G     0  3.9G   0% /run/shm
none            100M     0  100M   0% /run/user
/dev/xvdb        99G 1021M   93G   2% /concourse

# wait for worker to finish starting in other session

ubuntu@ip-172-31-33-124:~$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.9G   12K  3.9G   1% /dev
tmpfs           799M  348K  799M   1% /run
/dev/xvda1      7.8G  1.1G  6.3G  15% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
none            5.0M     0  5.0M   0% /run/lock
none            3.9G     0  3.9G   0% /run/shm
none            100M     0  100M   0% /run/user
/dev/xvdb        99G 1016M   93G   2% /concourse
/dev/loop0       99G  512K   97G   1% /concourse/volumes
```

Only difference I see is that your machine ended up with `/dev/mapper/concourse-data`. Maybe we don't correctly detect the size in that case?

**Comment by ipsi** [comment]
We've been seeing similar issues on our project - even with a 100GiB ephemeral disk, we seemed to wind up with at most 4.94GiB available for the docker container. `df -h` reported the following:

```
Filesystem      Size  Used Avail Use% Mounted on
udev            7.4G  4.0K  7.4G   1% /dev
tmpfs           1.5G  388K  1.5G   1% /run
/dev/xvda1       30G  1.3G   27G   5% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
none            5.0M     0  5.0M   0% /run/lock
none            7.4G     0  7.4G   0% /run/shm
none            100M     0  100M   0% /run/user
/dev/xvdb2       84G  2.9G   77G   4% /var/vcap/data
tmpfs           1.0M   20K 1004K   2% /var/vcap/data/sys/run
/dev/loop0       74G  891M   72G   2% /var/vcap/data/baggageclaim/volumes
```

Also, `/var/vcap/sys/log/fs_mounter.stdout.log` reported:

```
{"timestamp":"1473080376.915849447","source":"baggageclaim","message":"baggageclaim.run-command.ran","log_level":0,"data":{"args":["bash","-e","-x","-c","
		if [ ! -e $IMAGE_PATH ] || [ \"$(stat --printf=\"%s\" $IMAGE_PATH)\" != \"$SIZE_IN_BYTES\" ]; then
			touch $IMAGE_PATH
			truncate -s ${SIZE_IN_BYTES} $IMAGE_PATH
		fi

		lo=\"$(losetup -j $IMAGE_PATH | cut -d':' -f1)\"
		if [ -z \"$lo\" ]; then
			lo=\"$(losetup -f --show $IMAGE_PATH)\"
		fi

		if ! file $IMAGE_PATH | grep BTRFS; then
			/var/vcap/packages/btrfs_tools/sbin/mkfs.btrfs --nodiscard $IMAGE_PATH
		fi

		mkdir -p $MOUNT_PATH

		if ! mountpoint -q $MOUNT_PATH; then
			mount -t btrfs $lo $MOUNT_PATH
		fi
	"],"command":"/bin/bash","env":["PATH=/bin:/usr/bin:/sbin:/usr/sbin","MOUNT_PATH=/var/vcap/data/baggageclaim/volumes","IMAGE_PATH=/var/vcap/data/baggageclaim/volumes.img","SIZE_IN_BYTES=5304745984"],"session":"1","stderr":"+ '[' '!' -e /var/vcap/data/baggageclaim/volumes.img ']'
+ touch /var/vcap/data/baggageclaim/volumes.img
+ truncate -s 5304745984 /var/vcap/data/baggageclaim/volumes.img
++ losetup -j /var/vcap/data/baggageclaim/volumes.img
++ cut -d: -f1
+ lo=
+ '[' -z '' ']'
++ losetup -f --show /var/vcap/data/baggageclaim/volumes.img
+ lo=/dev/loop0
+ file /var/vcap/data/baggageclaim/volumes.img
+ grep BTRFS
+ /var/vcap/packages/btrfs_tools/sbin/mkfs.btrfs --nodiscard /var/vcap/data/baggageclaim/volumes.img
+ mkdir -p /var/vcap/data/baggageclaim/volumes
+ mountpoint -q /var/vcap/data/baggageclaim/volumes
+ mount -t btrfs /dev/loop0 /var/vcap/data/baggageclaim/volumes
","stdout":"btrfs-progs v4.4
See http://btrfs.wiki.kernel.org for more information.

Label:              (null)
UUID:               716eeb97-2f47-49d6-8434-4263abb1e9af
Node size:          16384
Sector size:        4096
Filesystem size:    4.94GiB
Block group profiles:
  Data:             single            8.00MiB
  Metadata:         DUP             260.94MiB
  System:           DUP              12.00MiB
SSD detected:       no
Incompat features:  extref, skinny-metadata
Number of devices:  1
Devices:
   ID        SIZE  PATH
    1     4.94GiB  /var/vcap/data/baggageclaim/volumes.img

"}}
```

(new lines and tabs normalised), compared the following (after changing to BTRFS)

```
{"timestamp":"1473085223.899114847","source":"baggageclaim","message":"baggageclaim.run-command.ran","log_level":0,"data":{"args":["bash","-e","-x","-c","
		if [ ! -e $IMAGE_PATH ] || [ \"$(stat --printf=\"%s\" $IMAGE_PATH)\" != \"$SIZE_IN_BYTES\" ]; then
			touch $IMAGE_PATH
			truncate -s ${SIZE_IN_BYTES} $IMAGE_PATH
		fi

		lo=\"$(losetup -j $IMAGE_PATH | cut -d':' -f1)\"
		if [ -z \"$lo\" ]; then
			lo=\"$(losetup -f --show $IMAGE_PATH)\"
		fi

		if ! file $IMAGE_PATH | grep BTRFS; then
			/var/vcap/packages/btrfs_tools/sbin/mkfs.btrfs --nodiscard $IMAGE_PATH
		fi

		mkdir -p $MOUNT_PATH

		if ! mountpoint -q $MOUNT_PATH; then
			mount -t btrfs $lo $MOUNT_PATH
		fi
	"],"command":"/bin/bash","env":["PATH=/bin:/usr/bin:/sbin:/usr/sbin","MOUNT_PATH=/var/vcap/data/baggageclaim/volumes","IMAGE_PATH=/var/vcap/data/baggageclaim/volumes.img","SIZE_IN_BYTES=79287025664"],"session":"1","stderr":"+ '[' '!' -e /var/vcap/data/baggageclaim/volumes.img ']'
+ touch /var/vcap/data/baggageclaim/volumes.img
+ truncate -s 79287025664 /var/vcap/data/baggageclaim/volumes.img
++ losetup -j /var/vcap/data/baggageclaim/volumes.img
++ cut -d: -f1
+ lo=
+ '[' -z '' ']'
++ losetup -f --show /var/vcap/data/baggageclaim/volumes.img
+ lo=/dev/loop0
+ file /var/vcap/data/baggageclaim/volumes.img
+ grep BTRFS
+ /var/vcap/packages/btrfs_tools/sbin/mkfs.btrfs --nodiscard /var/vcap/data/baggageclaim/volumes.img
+ mkdir -p /var/vcap/data/baggageclaim/volumes
+ mountpoint -q /var/vcap/data/baggageclaim/volumes
+ mount -t btrfs /dev/loop0 /var/vcap/data/baggageclaim/volumes
","stdout":"btrfs-progs v4.4
See http://btrfs.wiki.kernel.org for more information.

Label:              (null)
UUID:               a0c4d360-0be1-4ee0-8df1-26c55ab7cbf3
Node size:          16384
Sector size:        4096
Filesystem size:    73.84GiB
Block group profiles:
  Data:             single            8.00MiB
  Metadata:         DUP               1.01GiB
  System:           DUP              12.00MiB
SSD detected:       no
Incompat features:  extref, skinny-metadata
Number of devices:  1
Devices:
   ID        SIZE  PATH
    1    73.84GiB  /var/vcap/data/baggageclaim/volumes.img

"}}
```

As with @troyappcarousel, changing to `btrfs` seems to have fixed things for us.

We're running on AWS with the latest version of Concourse, BOSH, and BOSH Micro (AFAIK).

Stemcells:

```
+-----------------------------------------+---------------+---------+--------------------+
| Name                                    | OS            | Version | CID                |
+-----------------------------------------+---------------+---------+--------------------+
| bosh-aws-xen-hvm-ubuntu-trusty-go_agent | ubuntu-trusty | 3262.9* | ami-10d7a763 light |
+-----------------------------------------+---------------+---------+--------------------+
```

Releases:

```
+-------------+----------+-------------+
| Name        | Versions | Commit Hash |
+-------------+----------+-------------+
| concourse   | 2.0.2*   | 84ac63e     |
| garden-runc | 0.6.0*   | a77242dd    |
+-------------+----------+-------------+
```

**Comment by vito** [comment]
Seems to be this line that's the cause:

```
+ truncate -s 5304745984 /var/vcap/data/baggageclaim/volumes.img
```

5304745984 bytes = 4.9 GiB

vs.

```
+ truncate -s 79287025664 /var/vcap/data/baggageclaim/volumes.img
```

79287025664 = 73.8 GiB

It ended up just incorrectly sizing the image before. Not sure how or why.

**Comment by vito** [comment]
The bit that determines the size in the BOSH release is here:

https://github.com/concourse/concourse/blob/master/jobs/baggageclaim/templates/baggageclaim_ctl.erb#L36

**Comment by SillyMoo** [comment]
It seems to me somewhat unrealistic setting the loopback size to the size of the enclosing volume. Reason being work-dir is also used for storing images, etc. I wonder if it could be possible to specify a maximum size to avoid choking other users of that volume?

**Comment by billimek** [comment]
Not sure if this is in any way related, but we have been struggling with a similar issue on our ubuntu-based VMs running concourse via docker-compose.

The `/var/lib/docker/aufs/` tree would get very large for the containers representing the concourse worker process, e.g.:

```
47G  ./diff/1bb9{snip}/worker-state/volumes.img
47G  ./mnt/1bb9{snip}/worker-state/volumes.img
```

And would not clear unless we actually rebooted the server (restarting concourse, or blowing away the concourse containers completely and starting fresh, or even restarting the docker engine did not clear the usage).

We experimented with setting the `CONCOURSE_GARDEN_GRAPH_CLEANUP_THRESHOLD_IN_MEGABYTES` that gets passed to the concourse worker - and setting it to something smaller than the total size of the partition where docker writes the container volume.

Setting this seems to have made the problem go away, so far in two different concourse server instances where we were having nearly daily problems with the filesystem filling up.

**Comment by billimek** [comment]
I've been spending a lot of time looking at this from the client side and did some comparisons that may be useful:

## 2 servers, 1 test:

The hypothesis is that the regular server (ld5382) will see a constant growth of the worker /mnt/concourse filesystem with no reduction and the enhanced server (ld5458) will see usage reduce five minutes after a test completes

### configurations:
* **server1** : pristine concourse worker pointing to an ext4 filesystem (and associated volumes.img sparseimage) on /mnt/concourse - referenced as _current_
* **server2** : pristine concourse worker pointing to a real btrfs filesystem (not using a volumes.img sparseimage) on /mnt/concourse  - referenced as _new_

* flight-school test #1: run regular flight-school concourse job that completes successfully
* flight-school test #2: run the following flight-school concourse jobs in succession:
 * regular passing job using ruby docker image
 * failed job (using 'firefox' docker image  (in order to simulate consuming more space in failed jobs))
 * failed job (using the flow/java docker image (in order to simulate consuming more space in failed jobs))
 * regular passing job using ruby docker image

### results

#### flight-school test #1 (immediately after running)
* **_current_** : 2951Mb used in /mnt/concourse
* **_new_** : 1948Mb used in /mnt/concourse

#### flight-school test #1 (> 5 minutes after running - to allow for concourse worker to reap & clean-up)
* **_current_** : 2956Mb used in /mnt/concourse
* **_new_** : 1278Mb used in /mnt/concourse

#### flight-school test #2 (immediately after running)
* **_current_** : 8059Mb used in /mnt/concourse
* **_new_** : 5581Mb used in /mnt/concourse

#### flight-school test #2 (> 5 minutes after running - to allow for concourse worker to reap & clean-up)
* **_current_** : 8063Mb used in /mnt/concourse &  8 volumes, 2 containers
* **_new_** : 1295Mb used in /mnt/concourse & 8 volumes, 2 containers


**_current_ approach uses 522% more disk space than _new_ approach in this test**

**Comment by mrmcmuffinz** [comment]
Hi, 

I would like to add that I'm also seeing very similar behavior as well. I have docker setup with lvm storage for production and ever since I restarted my concourse instance this week via docker-compose I have noticed that the storage on our secondary drive gets sucked up:

```
root@acabrer-test-1-concourse-ci-1:~# df -h
Filesystem      Size  Used Avail Use% Mounted on
udev             16G     0   16G   0% /dev
tmpfs           3.2G  8.9M  3.2G   1% /run
/dev/xvda2       97G  2.7G   94G   3% /
tmpfs            16G     0   16G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs            16G     0   16G   0% /sys/fs/cgroup
/dev/xvda1      240M   97M  131M  43% /boot
tmpfs           3.2G     0  3.2G   0% /run/user/0
/dev/dm-3        10G  362M  9.7G   4% /var/lib/docker/devicemapper/mnt/2794d9346258990a989a3ddc945a2cc861d8ecba65525ff3c324124bb6f61eb9
shm              64M  4.0K   64M   1% /var/lib/docker/containers/bf65fc50000ed90627ec09f7373b2d25605720f4cf918f2c71a9e9f7a808c022/shm
/dev/dm-4        10G  477M  9.6G   5% /var/lib/docker/devicemapper/mnt/321cfdca9444b71fa82d13ee84f9cefa8454681968afba0597f2f42c39c763f0
shm              64M     0   64M   0% /var/lib/docker/containers/9bb314391776b07e02dc8e05f918122122ae5e678ebd5a4285167389ea2dad3c/shm
/dev/dm-5        10G   10G  512K 100% /var/lib/docker/devicemapper/mnt/5f1a42160f8deb46ed965b2f7949cf3aa720f193dec30dce1f247e72dfd4f190
shm              64M     0   64M   0% /var/lib/docker/containers/0c46057d9a2ed238fa8413252b1773503aca91998d0b914b68999c4e79086c12/shm
```

Upon inspection I found out who was eating it up:

```
root@acabrer-test-1-concourse-ci-1:/var/lib/docker/devicemapper/mnt/5f1a42160f8deb46ed965b2f7949cf3aa720f193dec30dce1f247e72dfd4f190/rootfs/worker-state# ls -lh
total 8.8G
drwxr-xr-x 3 root root  29 Jan 24 06:45 2.6.0
drwxr-xr-x 3 root root  50 Jan 24 07:04 depot
drwxr-xr-x 2 root root   6 Jan 24 06:45 volumes
-rw-r--r-- 1 root root 10G Jan 24 06:56 volumes.img
```

This causes all the pipeline to fail and not continue on because of lack of space to continue. Would appreciate some help on how to resolve this without having to increase my drives capacity. Right now we have a secondary drive of 100gb just for docker and that used to work perfectly, I'm wondering what is causing this issue now?

**Comment by mrmcmuffinz** [comment]
Update:

I was recommended to use `btrfs` for my storage setup with docker and by following the instructions provided by docker I believe my storage issues with concourse have gone away. Though I'm not entirely sure if this is dockers, storage or concourses fault. One thing I can say is that I was asked if I was using a docker image resource and the answer is yes.

**Comment by dmlemos** [comment]
I guess it is worth looking into alternatives like [LCFS](https://github.com/portworx/lcfs)

**Comment by vito** [comment]
This and confusion around it will be resolved by #1045 (switching off btrfs).
